### PR TITLE
Anonymous Struct literal `${a:1, b:2}`

### PR DESCRIPTION
**Pull request by ko1**
Try to introduce anonymous Struct literal.
https://bugs.ruby-lang.org/issues/16986

`${a:1, b:2}` is almost equal to `Struct.new(kw.keys).new(kw.values)` where `kw` is `{a:1, b:2}`.
Anonymous class is prepared at compile time.

### CODE REVIEW

File: compile.c [hunk 1]
```diff
@@ -4436,6 +4436,47 @@ compile_hash(rb_iseq_t *iseq, LINK_ANCHOR *const ret, const NODE *node, int popp
...
+    syms = ALLOC_N(VALUE, n);
+    list = node->nd_head;
+    while (list) {
+        NODE *key = list->nd_head;
+        syms[i++] = key->nd_lit;
+        list = list->nd_next;
+        NODE *val = list->nd_head;
+        list = list->nd_next;
+
+        CHECK(COMPILE_(ret, "struct value", val, popped));
```
Comment (nobu) [inline compile.c:4469]: When `val` has a syntax error, `syms` won't be freed.

Comment (ko1) [reply compile.c:4469]: Thank you. To debugging purpose, I changed it from `ALLOCA_N`, and I forget to restore it.

**Comment by ioquatix** [comment]
It's a fun idea.

I can see many usage.

**Comment by ys** [comment]
I always wanted something short like this, but do you think we could have something more explicit and verbose like `struct(a:1, b:2)`. Single symbols always end up with people googling around weirdly to find which one it is!

File: parse.y [hunk 2]
```diff
@@ -2893,6 +2897,14 @@ primary		: literal
 		    /*% %*/
 		    /*% ripper: hash!(escape_Qundef($2)) %*/
 		    }
+                | '$' tLBRACE kw_assoc_list '}'
+                    {
+		    /*%%%*/
+                        $$ = new_struct(p, $3, &@$);
+                        $$->nd_brace = TRUE;
```
Comment (nobu) [inline parse.y:2904]: This seems unnecessary, because it is not a `NODE_HASH`.

**Comment by inopinatus** [comment]
It's a nice idea. I can see myself eval'ing these instead of using an OpenStruct, since structs are faster.

Can I suggest that the output of `Struct#inspect` change to also be the literal syntax? Currently:

    foo = ${a:42, b:99}
    foo.inspect #=> "#<struct a=42, b=99>"

instead:

    foo = ${a:1, b:2}
    foo.inspect #=> "#{a: 1, b: 2}"
    eval foo.inspect #=> same

Look at this also gives me an idea. What if `${}` is a general purpose anonymous constructor, that defaults to Struct? So that

    ${a: 42, b:99}

produces the expected instance of an anonymous struct, but

    ${Coordinate x: 7, y:24}

is equivalent to either the class-builder pattern of:

    Coordinate.new(:x, :y).new(7, 24)

or possibly module composition into Struct:

   Struct.new(:x, :y).new(7, 24).extend(Coordinate)

**Comment by ko1** [comment]
DO NOT discuss spec here. Please write a comment on redmine ticket.

**Comment by pynixwang** [comment]
just delegate `method_missing` to `[]`, then you can use `.` syntax

**Pull request closed without merging**